hadoop-hdds/docs/content/design/leader-execution/leader-execution.md

```diff
@@ -0,0 +1,422 @@
+Message format
+```
+message PersistDbRequest {
+  repeated PersistDbInfo info = 1;
+}
+
+message PersistDbInfo {
+  optional int64 index = 1;
+  repeated DBTableUpdate tableUpdates = 2;
+  repeated DBTableMove tableMoves = 3;
+  repeated DBBucketQuotaUpdate bucketQuotaUpdate = 4;
+}
+
+message DBBucketQuotaUpdate {
+required string volName = 1;
+required string bucketName = 2;
+required int64 diffUsedBytes = 3;
+required int64 diffUsedNamespace = 4;
+optional uint64 bucketObjectId = 5;
+}
+message DBTableMove {
+  optional string srcTableName = 1;
+  optional string destTableName = 2;
+  optional string srcKey = 3;
+  optional string destKey = 4;
+}
+message DBTableUpdate {
+  required string tableName = 1;
+  repeated DBTableRecord records = 2;
+}
+message DBTableRecord {
+  required bytes key = 1;
+  optional bytes value = 2;
+}
+```
+
+### Replay of client request handling
+
+In old flow, request is submitted to ratis directly with ClientId and CallId (From HadoopRPC callback), and same used by ratis to handle request replay (or retry).
+If ratis finds the another request with same ClientId and CallId, it returns response of last request itself matching.
+Additionally, mapping of ClientId and CallId is present in memory and gets flushed out on restart, so it does not have consistent behaviour for handling same.
+
+For new flow, since only db changes are added to ratis, so this mechanism can not be used. To handle this, need to have similar mechanism, and improve over persistence for replay.
+
+`Client Request --> Gatekeeper --> check for request exist in cache for ClientId and CallId
+--> If exist, return cached response
+--> Else continue request handling`
+
+#### Client request replay at leader node
+- When request is received at leader node, it will cache the request in replayCache immediately
+- When request is received again with same ClientId and CallId, it will check replayCache and if entry exist,
+    - If response in cache is not available (request handling in progess), wait for availability and timeout - 60 sec
+    - If response available, return immediately
+- If entry does not exist, it will process handling request normally
+
+
+#### Replay cache distribution to other nodes
+Request - response will be cached to other node via ratis distribution
+- It will be added to memory cache with expiry handling
+- Also will be added to DB for persistence for restart handing
+
+Below information will be sync to all nodes via ratis:
+```
+message ClientRequestInfo {
+  optional string uuidClientId = 1;
+  optional uint64 callId = 2;
+  optional unint64 timestamp = 5;
+  optional OMResponse response = 3;
+}
+```
+
+Refer [Execution Control Message](leader-execution.md#control-request) for format.
+
+DB Persistence:
+```
+New rocksDB table: ClientRequestReplyTable
+Key: String: ClientId#CallId
+Value: ClientRequestInfo
+```
+
+#### Memory caching:
+```
+Memory Map: ClientId#CallId Vs Response
+Expiry: 10 minute (as current default for ratis)
```

Comment:
This expiry is done independently for each OM node? It won't be replicated from leader?

Reply:
Leader will sync below information with db changes,
- callI and clientId (used to identify retry of request)
- timestamp
- response message (to be replied back if same request is retried as ratis behavior)

Each node will make use of above information once received in applyTransaction and handle independently.